File: evidence/v1_1/ai_eval/subjects/counted/r1/tally/cpp/main.cpp
// Tallies the values above a threshold on k threads, one contiguous chunk each.
#include <cstdint>
#include <iostream>
#include <thread>
#include <vector>

struct Tally {
  std::uint64_t above = 0, sum = 0, max = 0;
};

static void tally_chunk(const std::vector<std::uint64_t>& values, std::size_t lo, std::size_t hi, std::uint64_t t, Tally& out) {
  for (std::size_t i = lo; i < hi; ++i) {
    if (values[i] > t) {
      out.above += 1;
      out.sum += values[i];  // unsigned: wraps modulo 2^64
    }
    if (values[i] > out.max) out.max = values[i];
  }
}

int main() {
  std::ios::sync_with_stdio(false);
  std::size_t k = 0, n = 0;
  std::uint64_t t = 0;
  if (!(std::cin >> k >> n >> t)) return 1;
  std::vector<std::uint64_t> values(n);
  for (auto& v : values) std::cin >> v;

  std::vector<Tally> partials(k);
  std::vector<std::thread> threads;
  for (std::size_t j = 0; j < k; ++j) {
    std::size_t lo = n * j / k, hi = n * (j + 1) / k;
    threads.emplace_back([&, lo, hi, j] { tally_chunk(values, lo, hi, t, partials[j]); });
  }
  for (auto& th : threads) th.join();

  Tally total;
  for (const auto& p : partials) {
    total.above += p.above;
    total.sum += p.sum;
    if (p.max > total.max) total.max = p.max;
  }

  std::cout << "above " << total.above << "\nsum " << total.sum << '\n';
  if (n == 0) std::cout << "max none\n";
  else std::cout << "max " << total.max << '\n';
  return 0;
}

File: evidence/v1_1/ai_eval/subjects/counted/r1/tally/rust/main.rs
// Tallies the values above a threshold on k threads, one contiguous chunk each.
use std::io::{self, Read};
use std::thread;

#[derive(Clone, Copy)]
struct Tally {
    above: u64,
    sum: u64,
    max: u64,
}

fn tally_chunk(chunk: &[u64], t: u64) -> Tally {
    let mut total = Tally { above: 0, sum: 0, max: 0 };
    for &v in chunk {
        if v > t {
            total.above += 1;
            total.sum = total.sum.wrapping_add(v);
        }
        total.max = total.max.max(v);
    }
    total
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut tokens = input.split_ascii_whitespace();
    let k: usize = tokens.next().unwrap().parse().unwrap();
    let n: usize = tokens.next().unwrap().parse().unwrap();
    let t: u64 = tokens.next().unwrap().parse().unwrap();
    let values: Vec<u64> = tokens.take(n).map(|v| v.parse().unwrap()).collect();

    let totals = thread::scope(|s| {
        let handles: Vec<_> = (0..k)
            .map(|j| {
                let chunk = &values[n * j / k..n * (j + 1) / k];
                s.spawn(move || tally_chunk(chunk, t))
            })
            .collect();
        handles.into_iter().map(|h| h.join().unwrap()).collect::<Vec<_>>()
    });

    let total = totals.into_iter().fold(
        Tally { above: 0, sum: 0, max: 0 },
        |acc, part| Tally {
            above: acc.above + part.above,
            sum: acc.sum.wrapping_add(part.sum),
            max: acc.max.max(part.max),
        },
    );
    println!("above {}\nsum {}", total.above, total.sum);
    if n == 0 {
        println!("max none");
    } else {
        println!("max {}", total.max);
    }
}
